sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PartitionPruning.scala

```diff
@@ -0,0 +1,259 @@
+    case _: In | _: InSet => true
+    case _: StringPredicate => true
+    case _ => false
+  }
+
+  /**
+   * Search a filtering predicate in a given logical plan
+   */
+  private def hasSelectivePredicate(plan: LogicalPlan): Boolean = {
+    plan.find {
+      case f: Filter => isLikelySelective(f.condition)
+      case _ => false
+    }.isDefined
+  }
+
+  /**
+   * To be able to prune partitions on a join key, the filtering side needs to
+   * meet the following requirements:
+   *   (1) it can not be a stream
+   *   (2) it needs to contain a selective predicate used for filtering
+   */
+  private def hasPartitionPruningFilter(plan: LogicalPlan): Boolean = {
+    !plan.isStreaming && hasSelectivePredicate(plan)
+  }
+
+  private def canPruneLeft(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | RightOuter => true
+    case _ => false
+  }
+
+  private def canPruneRight(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftOuter => true
+    case _ => false
+  }
+
+  private def prune(plan: LogicalPlan): LogicalPlan = {
+    plan transformUp {
+      // skip this rule if there's already a DPP subquery on the LHS of a join
+      case j @ Join(Filter(_: DynamicPruningSubquery, _), _, _, _, _) => j
+      case j @ Join(_, Filter(_: DynamicPruningSubquery, _), _, _, _) => j
+      case j @ Join(left, right, joinType, Some(condition), hint) =>
+        var newLeft = left
+        var newRight = right
+
+        // extract the left and right keys of the join condition
+        val (leftKeys, rightKeys) = j match {
+          case ExtractEquiJoinKeys(_, lkeys, rkeys, _, _, _, _) => (lkeys, rkeys)
+          case _ => (Nil, Nil)
+        }
+
+        // checks if two expressions are on opposite sides of the join
+        def fromDifferentSides(x: Expression, y: Expression): Boolean = {
+          def fromLeftRight(x: Expression, y: Expression) =
+            !x.references.isEmpty && x.references.subsetOf(left.outputSet) &&
+              !y.references.isEmpty && y.references.subsetOf(right.outputSet)
+          fromLeftRight(x, y) || fromLeftRight(y, x)
+        }
+
+        splitConjunctivePredicates(condition).foreach {
+          case EqualTo(a: Expression, b: Expression)
+              if fromDifferentSides(a, b) =>
+            val (l, r) = if (a.references.subsetOf(left.outputSet) &&
+              b.references.subsetOf(right.outputSet)) {
+              a -> b
+            } else {
+              b -> a
+            }
+
+            // there should be a partitioned table and a filter on the dimension table,
+            // otherwise the pruning will not trigger
+            var partScan = getPartitionTableScan(l, left)
+            if (partScan.isDefined && canPruneLeft(joinType) &&
+                hasPartitionPruningFilter(right)) {
+              val hasBenefit = pruningHasBenefit(l, partScan.get, r, right)
+              newLeft = insertPredicate(l, newLeft, r, right, rightKeys, hasBenefit)
```

Comment:
Recursion will occur here. When multiple left joins meet the condition, the plan will be amplified.